pandas/core/internals/blocks.py

```diff
@@ -489,17 +513,19 @@ def fillna(
             # we can't process the value, but nothing to do
             return [self] if inplace else [self.copy()]
 
-        # operate column-by-column
-        def f(mask, val, idx):
-            block = self.coerce_to_target_dtype(value)
-
-            # slice out our block
-            if idx is not None:
-                # i.e. self.ndim == 2
-                block = block.getitem_block(slice(idx, idx + 1))
-            return block.fillna(value, limit=limit, inplace=inplace, downcast=None)
+        elif self.ndim == 1 or self.shape[0] == 1:
+            blk = self.coerce_to_target_dtype(value)
+            # bc we have already cast, inplace=True may avoid an extra copy
+            return blk.fillna(value, limit=limit, inplace=True, downcast=None)
 
-        return self.split_and_operate(None, f, inplace)
+        else:
+            # operate column-by-column
+            res_blocks = []
+            nbs = self._split()
```

Comment:
if you can remove _split entirely and *just* use the decorator, I think that might be better.

Reply:
trouble is we also use it in BlockManager.get_bool_data

Reply:
ohh, i think i may now understand the confusion.

is the confusing thing "why are we iterating over self._split() here instead of using the decorator?  i.e. why not use the decorator everywhere?"?

if so, the reason is bc in this method, the conditions under which we want to split are slightly different from the conditions checked in the decorator.  im hopeful we can get some further simplification, but dont want to get it by jamming too much complexity into maybe_split (which is what got split_and_operate confusing in the first place)